ts/Core/Series/Series.ts

```diff
@@ -1389,23 +1388,33 @@ class Series {
                 (series as any)[key + 'Data'].length = 0;
             });
 
-            // In turbo mode, only one- or twodimensional arrays of numbers
-            // are allowed. The first value is tested, and we assume that
-            // all the rest are defined the same way. Although the 'for'
-            // loops are similar, they are repeated inside each if-else
-            // conditional for max performance.
-            if (turboThreshold && dataLength > turboThreshold) {
-
-                firstPoint = series.getFirstValidPoint(data);
+            // In turbo mode, look for one- or twodimensional arrays of numbers.
+            // The first and the last valid value are tested, and we assume that
+            // all the rest are defined the same way. Although the 'for' loops
+            // are similar, they are repeated inside each if-else conditional
+            // for max performance.
+            let runTurbo = turboThreshold && dataLength > turboThreshold;
+            if (runTurbo) {
+
+                const firstPoint = series.getFirstValidPoint(data),
+                    lastPoint = series.getFirstValidPoint(
+                        data, dataLength - 1, -1
+                    ),
+                    isNumericArray = (a: unknown): a is Array<number> =>
```

Comment:
Based on the names, assigned type, and docs wording the only allowed format for the data points, when given as an array, is an array with numbers only. The current code works without this limitation. Keys could be used to provide other types to the data point array and even the first element doesn't have to be a number when `keys` give `x` and `y` other indexes (and even that is not required for some cases). [Demo](https://jsfiddle.net/BlackLabel/sy16t5L2/). The currently proposed change might be a breaking change for some use cases.

Reply:
Good point. I introduced that limitation because `[name, y]` tuples wouldn't work without it: https://jsfiddle.net/highcharts/Lswhug81/ . I also assume this will be the case for the [upcoming date strings](https://github.com/highcharts/highcharts/pull/21258).

But I'll do some more digging and see how `keys` play in.

> The currently proposed change might be a breaking change for some use cases.
I don't think so, because when `runTurbo` goes false, we go into the `applyOptions` approach, which may be a bit slower but works with all types of configurations.